File: bindings/dotnet/dotnet/StateSet/StateSetCommerce.cs
using System.Runtime.InteropServices;
using System.Text.Json;

namespace StateSet.Embedded;

/// <summary>
/// StateSet Embedded Commerce - The SQLite of Commerce
///
/// A zero-dependency, local-first commerce engine for .NET applications.
/// </summary>
/// <example>
/// <code>
/// using var commerce = new StateSetCommerce("store.db");
///
/// var customer = commerce.Customers.Create(
///     email: "alice@example.com",
///     firstName: "Alice",
///     lastName: "Smith"
/// );
///
/// var product = commerce.Products.Create(
///     name: "Premium Widget",
///     sku: "WIDGET-001",
///     price: 29.99m
/// );
///
/// var order = commerce.Orders.Create(
///     customerId: customer.Id,
///     items: new[] { new OrderItem { Sku = "WIDGET-001", Name = "Widget", Quantity = 2, UnitPrice = 29.99 } },
///     currency: "USD"
/// );
/// </code>
/// </example>
public sealed class StateSetCommerce : IDisposable
{
    private IntPtr _handle;
    private bool _disposed;
    internal StateSetDotNetStore Store { get; } = new();

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower
    };

    /// <summary>
    /// Customers API
    /// </summary>
    public CustomersApi Customers { get; }

    /// <summary>
    /// Products API
    /// </summary>
    public ProductsApi Products { get; }

    /// <summary>
    /// Orders API
    /// </summary>
    public OrdersApi Orders { get; }

    /// <summary>
    /// Inventory API
    /// </summary>
    public InventoryApi Inventory { get; }

    /// <summary>
    /// Carts API
    /// </summary>
    public CartsApi Carts { get; }

    /// <summary>
    /// Returns API
    /// </summary>
    public ReturnsApi Returns { get; }

    /// <summary>
    /// Payments API
    /// </summary>
    public PaymentsApi Payments { get; }

    /// <summary>
    /// Analytics API
    /// </summary>
    public AnalyticsApi Analytics { get; }

    /// <summary>
    /// Shipments API
    /// </summary>
    public ShipmentsApi Shipments { get; }

    /// <summary>
    /// Warranties API
    /// </summary>
    public WarrantiesApi Warranties { get; }

    /// <summary>
    /// Suppliers API
    /// </summary>
    public SuppliersApi Suppliers { get; }

    /// <summary>
    /// Purchase Orders API
    /// </summary>
    public PurchaseOrdersApi PurchaseOrders { get; }

    /// <summary>
    /// Invoices API
    /// </summary>
    public InvoicesApi Invoices { get; }

    /// <summary>
    /// Bill of Materials API
    /// </summary>
    public BomApi Bom { get; }

    /// <summary>
    /// Work Orders API
    /// </summary>
    public WorkOrdersApi WorkOrders { get; }

    /// <summary>
    /// Currency API
    /// </summary>
    public CurrencyApi Currency { get; }

    /// <summary>
    /// Subscriptions API
    /// </summary>
    public SubscriptionsApi Subscriptions { get; }

    /// <summary>
    /// Promotions API
    /// </summary>
    public PromotionsApi Promotions { get; }

    /// <summary>
    /// Tax API
    /// </summary>
    public TaxApi Tax { get; }

    /// <summary>
    /// Quality API
    /// </summary>
    public QualityApi Quality { get; }

    /// <summary>
    /// Lots API
    /// </summary>
    public LotsApi Lots { get; }

    /// <summary>
    /// Serials API
    /// </summary>
    public SerialsApi Serials { get; }

    /// <summary>
    /// Warehouse API
    /// </summary>
    public WarehouseApi Warehouse { get; }

    /// <summary>
    /// Receiving API
    /// </summary>
    public ReceivingApi Receiving { get; }

    /// <summary>
    /// Fulfillment API
    /// </summary>
    public FulfillmentApi Fulfillment { get; }

    /// <summary>
    /// Accounts Payable API
    /// </summary>
    public AccountsPayableApi AccountsPayable { get; }

    /// <summary>
    /// Accounts Receivable API
    /// </summary>
    public AccountsReceivableApi AccountsReceivable { get; }

    /// <summary>
    /// Cost Accounting API
    /// </summary>
    public CostAccountingApi CostAccounting { get; }

    /// <summary>
    /// Credit API
    /// </summary>
    public CreditApi Credit { get; }

    /// <summary>
    /// Backorders API
    /// </summary>
    public BackordersApi Backorders { get; }

    /// <summary>
    /// General Ledger API
    /// </summary>
    public GeneralLedgerApi GeneralLedger { get; }

    /// <summary>
    /// Create a new Commerce instance
    /// </summary>
    /// <param name="dbPath">Path to SQLite database file, or ":memory:" for in-memory database</param>
    /// <exception cref="StateSetException">Thrown if database initialization fails</exception>
    public StateSetCommerce(string dbPath)
    {
        _handle = NativeMethods.stateset_commerce_new(dbPath);
        if (_handle == IntPtr.Zero)
        {
            throw new StateSetException("Failed to create commerce instance");
        }

        Customers = new CustomersApi(this);
        Products = new ProductsApi(this);
        Orders = new OrdersApi(this);
        Inventory = new InventoryApi(this);
        Carts = new CartsApi(this);
        Returns = new ReturnsApi(this);
        Payments = new PaymentsApi(this);
        Analytics = new AnalyticsApi(this);
        Shipments = new ShipmentsApi(this);
        Warranties = new WarrantiesApi(this);
        Suppliers = new SuppliersApi(this);
        PurchaseOrders = new PurchaseOrdersApi(this);
        Invoices = new InvoicesApi(this);
        Bom = new BomApi(this);
        WorkOrders = new WorkOrdersApi(this);
        Currency = new CurrencyApi(this);
        Subscriptions = new SubscriptionsApi(this);
        Promotions = new PromotionsApi(this);
        Tax = new TaxApi(this);
        Quality = new QualityApi(this);
        Lots = new LotsApi(this);
        Serials = new SerialsApi(this);
        Warehouse = new WarehouseApi(this);
        Receiving = new ReceivingApi(this);
        Fulfillment = new FulfillmentApi(this);
        AccountsPayable = new AccountsPayableApi(this);
        AccountsReceivable = new AccountsReceivableApi(this);
        CostAccounting = new CostAccountingApi(this);
        Credit = new CreditApi(this);
        Backorders = new BackordersApi(this);
        GeneralLedger = new GeneralLedgerApi(this);
    }

    internal IntPtr Handle
    {
        get
        {
            ObjectDisposedException.ThrowIf(_disposed, this);
            return _handle;
        }
    }

    internal static T? ParseJson<T>(IntPtr ptr) where T : class
    {
        if (ptr == IntPtr.Zero)
            return null;

        try
        {
            var json = Marshal.PtrToStringUTF8(ptr);
            if (string.IsNullOrEmpty(json))
                return null;

            return JsonSerializer.Deserialize<T>(json, JsonOptions);
        }
        finally
        {
            NativeMethods.stateset_string_free(ptr);
        }
    }

    internal static T ParseJsonRequired<T>(IntPtr ptr) where T : class
    {
        var result = ParseJson<T>(ptr);
        return result ?? throw new StateSetException("Failed to parse response");
    }

    internal static List<T> ParseJsonList<T>(IntPtr ptr) where T : class
    {
        if (ptr == IntPtr.Zero)
            return new List<T>();

        try
        {
            var json = Marshal.PtrToStringUTF8(ptr);
            if (string.IsNullOrEmpty(json))
                return new List<T>();

            return JsonSerializer.Deserialize<List<T>>(json, JsonOptions) ?? new List<T>();
        }
        finally
        {
            NativeMethods.stateset_string_free(ptr);
        }
    }

    /// <summary>
    /// Dispose of the commerce instance and release native resources
    /// </summary>
    public void Dispose()
    {
        if (_disposed)
            return;

        if (_handle != IntPtr.Zero)
        {
            NativeMethods.stateset_commerce_free(_handle);
            _handle = IntPtr.Zero;
        }

        _disposed = true;
    }
}

/// <summary>
/// Exception thrown when a StateSet operation fails
/// </summary>
public class StateSetException : Exception
{
    public StateSetException(string message) : base(message) { }
    public StateSetException(string message, Exception inner) : base(message, inner) { }
}

File: bindings/swift/Sources/StateSet/StateSetCommerce.swift
import Foundation
import StateSetC

/// StateSet Embedded Commerce - The SQLite of Commerce
///
/// A zero-dependency, local-first commerce engine for Swift/iOS applications.
///
/// Example usage:
/// ```swift
/// let commerce = try StateSetCommerce(dbPath: "store.db")
///
/// let customer = try commerce.customers.create(
///     email: "alice@example.com",
///     firstName: "Alice",
///     lastName: "Smith"
/// )
///
/// let product = try commerce.products.create(
///     name: "Premium Widget",
///     sku: "WIDGET-001",
///     price: 29.99
/// )
///
/// let order = try commerce.orders.create(
///     customerId: customer.id,
///     items: [OrderItem(sku: "WIDGET-001", name: "Widget", quantity: 2, unitPrice: 29.99)],
///     currency: "USD"
/// )
/// ```
public final class StateSetCommerce {
    private var handle: StateSetHandle?

    /// Customers API
    public private(set) lazy var customers = CustomersAPI(commerce: self)

    /// Products API
    public private(set) lazy var products = ProductsAPI(commerce: self)

    /// Orders API
    public private(set) lazy var orders = OrdersAPI(commerce: self)

    /// Inventory API
    public private(set) lazy var inventory = InventoryAPI(commerce: self)

    /// Carts API
    public private(set) lazy var carts = CartsAPI(commerce: self)

    /// Returns API
    public private(set) lazy var returns = ReturnsAPI(commerce: self)

    /// Payments API
    public private(set) lazy var payments = PaymentsAPI(commerce: self)

    /// Analytics API
    public private(set) lazy var analytics = AnalyticsAPI(commerce: self)

    /// Shipments API
    public private(set) lazy var shipments = ShipmentsAPI(commerce: self)

    /// Warranties API
    public private(set) lazy var warranties = WarrantiesAPI(commerce: self)

    /// Suppliers API
    public private(set) lazy var suppliers = SuppliersAPI(commerce: self)

    /// Purchase Orders API
    public private(set) lazy var purchaseOrders = PurchaseOrdersAPI(commerce: self)

    /// Invoices API
    public private(set) lazy var invoices = InvoicesAPI(commerce: self)

    /// Bill of Materials API
    public private(set) lazy var bom = BOMAPI(commerce: self)

    /// Work Orders API
    public private(set) lazy var workOrders = WorkOrdersAPI(commerce: self)

    /// Currency API
    public private(set) lazy var currency = CurrencyAPI(commerce: self)

    /// Subscriptions API
    public private(set) lazy var subscriptions = SubscriptionsAPI(commerce: self)

    /// Promotions API
    public private(set) lazy var promotions = PromotionsAPI(commerce: self)

    /// Tax API
    public private(set) lazy var tax = TaxAPI(commerce: self)

    /// Quality API
    public private(set) lazy var quality = QualityAPI(commerce: self)

    /// Lots API
    public private(set) lazy var lots = LotsAPI(commerce: self)

    /// Serials API
    public private(set) lazy var serials = SerialsAPI(commerce: self)

    /// Warehouse API
    public private(set) lazy var warehouse = WarehouseAPI(commerce: self)

    /// Receiving API
    public private(set) lazy var receiving = ReceivingAPI(commerce: self)

    /// Fulfillment API
    public private(set) lazy var fulfillment = FulfillmentAPI(commerce: self)

    /// Accounts Payable API
    public private(set) lazy var accountsPayable = AccountsPayableAPI(commerce: self)

    /// Accounts Receivable API
    public private(set) lazy var accountsReceivable = AccountsReceivableAPI(commerce: self)

    /// Cost Accounting API
    public private(set) lazy var costAccounting = CostAccountingAPI(commerce: self)

    /// Credit API
    public private(set) lazy var credit = CreditAPI(commerce: self)

    /// Backorders API
    public private(set) lazy var backorders = BackordersAPI(commerce: self)

    /// General Ledger API
    public private(set) lazy var generalLedger = GeneralLedgerAPI(commerce: self)

    /// Create a new Commerce instance
    /// - Parameter dbPath: Path to SQLite database file, or ":memory:" for in-memory database
    /// - Throws: StateSetError if database initialization fails
    public init(dbPath: String) throws {
        handle = stateset_commerce_new(dbPath)
        if handle == nil {
            throw StateSetError.initializationFailed("Failed to create commerce instance")
        }
    }

    deinit {
        close()
    }

    /// Close the commerce instance and release native resources
    public func close() {
        if let h = handle {
            stateset_commerce_free(h)
            handle = nil
        }
    }

    internal func getHandle() throws -> StateSetHandle {
        guard let h = handle else {
            throw StateSetError.invalidHandle
        }
        return h
    }

    internal func parseJSON<T: Decodable>(_ ptr: UnsafeMutablePointer<CChar>?) throws -> T {
        guard let ptr = ptr else {
            throw StateSetError.nullPointer
        }
        defer { stateset_string_free(ptr) }

        let jsonString = String(cString: ptr)
        guard let data = jsonString.data(using: .utf8) else {
            throw StateSetError.invalidJSON("Failed to convert string to data")
        }

        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return try decoder.decode(T.self, from: data)
    }

    internal func parseOptionalJSON<T: Decodable>(_ ptr: UnsafeMutablePointer<CChar>?) -> T? {
        guard let ptr = ptr else { return nil }
        defer { stateset_string_free(ptr) }

        let jsonString = String(cString: ptr)
        guard let data = jsonString.data(using: .utf8) else { return nil }

        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return try? decoder.decode(T.self, from: data)
    }
}

/// StateSet error types
public enum StateSetError: Error, LocalizedError {
    case initializationFailed(String)
    case invalidHandle
    case nullPointer
    case invalidJSON(String)
    case invalidUUID
    case operationFailed(String)

    public var errorDescription: String? {
        switch self {
        case .initializationFailed(let msg): return "Initialization failed: \(msg)"
        case .invalidHandle: return "Invalid commerce handle"
        case .nullPointer: return "Null pointer returned from native code"
        case .invalidJSON(let msg): return "Invalid JSON: \(msg)"
        case .invalidUUID: return "Invalid UUID format"
        case .operationFailed(let msg): return "Operation failed: \(msg)"
        }
    }
}
